test/ts/endToEndLibraryToContract.test.ts

```diff
@@ -0,0 +1,420 @@
+import "mocha";
+import * as chai from "chai";
+import { solidity, loadFixture } from "ethereum-waffle";
+import { BigNumber, defaultAbiCoder } from "ethers/utils";
+import { fnIt } from "@pisa-research/test-utils";
+import {
+  RelayHubFactory,
+  MsgSenderExampleFactory,
+  RelayHub,
+  MultiNonceReplayProtection,
+  BitFlipReplayProtection,
+  RelayHubForwarder,
+  ProxyAccountDeployerFactory,
+  ProxyAccount,
+  ProxyAccountForwarder,
+  ProxyAccountFactory,
+} from "../../src";
+import { Provider } from "ethers/providers";
+import { Wallet } from "ethers/wallet";
+import { ChainID } from "../../src/ts/forwarders/forwarderFactory";
+import { flipBit } from "../utils/bitflip-utils";
+
+const expect = chai.expect;
+chai.use(solidity);
+
+let relayHubType: RelayHub;
+type relayHubFunctions = typeof relayHubType.functions;
+
+let proxyAccount: ProxyAccount;
+type proxyAccountFunctions = typeof proxyAccount.functions;
+
+async function setup(provider: Provider, [admin, owner, sender]: Wallet[]) {
+  const relayHub = await new RelayHubFactory(admin).deploy();
+
+  const proxyDeployerFactory = new ProxyAccountDeployerFactory(admin);
+  const proxyDeployerCreationTx = proxyDeployerFactory.getDeployTransaction();
+
+  const proxyDeployerCreation = await admin.sendTransaction(
+    proxyDeployerCreationTx
+  );
+  const proxyResult = await proxyDeployerCreation.wait(1);
+
+  const proxyDeployer = proxyDeployerFactory.attach(
+    proxyResult.contractAddress!
+  );
+  // when(PROXY_ACCOUNT_DEPLOYER_ADDRESS).thenReturn(proxyDeployer.address);
+
+  return {
+    provider,
+    relayHub,
+    proxyDeployer,
+    owner,
+    sender,
+  };
+}
+
+describe("End to End Library to Contract", () => {
+  fnIt<relayHubFunctions>(
+    (a) => a.forward,
+    "reset the forwarder 5 times and each time send 10 transactions. setting: relayhub & bitflip protection",
+    async () => {
+      const { relayHub, owner, sender } = await loadFixture(setup);
+
+      // Reset the forwarder 5 times
+      for (let j = 0; j < 5; j++) {
```

Comment:
Why is the same test repeated many times? What's the difference?
Same for other tests below.

Reply:
so they are testing different settings. e.g. relayhub + bitflip. proxyhub + nonce.

Reply:
I mean, why 5 times? `j` is not used in the test, so what's changing between different iterations of the loop?